BUG: DataFrame.diff(axis=0) with DatetimeTZ data

mroeschke opened the pull request:
- [x] closes #18578
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Unfortunately, `DataFrame.diff` doesn't work correctly with axis=1; it returns values like so after this patch:

```
In [2]:         df = DataFrame({0: date_range('2010', freq='D', periods=2, tz='utc'),
   ...:                         1: date_range('2010', freq='D', periods=2, tz='utc')})
   ...:

In [3]: df.diff(axis=1)
Out[3]:
       0      1
0    NaT    NaT
1    NaT    NaT
```

I believe this is because `diff` is implemented by iterating over the 2 DatetimeTZBlocks (via `BlockManager.apply`, but `diff` needs data from both blocks at once. (Is it possible to consolidate 2 DatetimeTZBlocks?) I made this raise a `NotImplimentedError` for now.

mroeschke committed:
BUG: DataFrame.diff(axis=0) with DatetimeTZ data

add whatsnew

clarify comment

Add addtional tests

move diff into its own function in DatetimeTZBlock

Use correct placement

fix failing test

formatting

jreback commented on pandas/core/algorithms.py:
```diff
@@ -1499,7 +1499,12 @@ def diff(arr, n, axis=0):
     if needs_i8_conversion(arr):
         dtype = np.float64
-        arr = arr.view('i8')
+        if is_datetime64tz_dtype(arr):
```
so I would leave this alone.

jreback commented on pandas/core/internals.py:
```diff
@@ -1269,6 +1269,11 @@ def take_nd(self, indexer, axis, new_mgr_locs=None, fill_tuple=None):
         """ return block for the diff of the values """
+        if isinstance(self, DatetimeTZBlock) and axis == 0:
+            # This method will iterate (self.apply) over each DateTimeTZBlock
+            # but axis=0 will need to gather data from all DateTimeTZBlocks
```
this is not a good long term soln, but you can put a diff method in DatetimeTZBlock to handle this, kind of like we do for shift. Its not great, but this whole diff / shift thing is pretty fundamentally broken ATM. The routines are not general enough to graciously handle the various types of inputs (e.g. 2-D arrays, 1-D arrays, 1-D non-arrays). Further algos.diff hasn't seen much attention to handle tz-aware things. E.g. you have to have a uniform tz in order to diff in the first place.

jreback commented on pandas/tests/frame/test_timeseries.py:
```diff
@@ -57,6 +57,22 @@ def test_diff(self):
 
+    @pytest.mark.parametrize('axis', [0, 1])
+    @pytest.mark.parametrize('tz', [None, 'UTC'])
+    @pytest.mark.xfail(raises=NotImplementedError)
```
putting the xfail here means *everything* xfails, so this is not a very useful test

mroeschke commented:
I moved diff to it's own function in `DatetimeTZBlock`. I am not to familiar with this section of the codebase (I believe I should be returning a `TimedeltaBlock` but I am not familiar with what should be passed to `placement`), so a test I added is going to fail. 

`df.shift(axis=1)` for datetimetz data wasn't working correctly as well, so I am raising a `NotImplementedError` for now.

mroeschke commented:
Any insight on the use of `placement` when constructing a new `Timedelta` block from the result would be much appreciated @jreback. 

I am currently failing the check, `len(self.mgr_locs) != len(self.values)`, when constructing a new block. I am passing in the same `mgr_locs` from the prior frame (where the data was datetimetz data), but I am not sure if that's entirely correct if the new frame should have a `Timedelta` block.

TomAugspurger commented:
> I am currently failing the check, len(self.mgr_locs) != len(self.values), when constructing a new block.

I think the issue is that `DatetimeTZBlock` inherits NonConsolidatableMixin,  while `TimedeltaBlock` doesn't, so a couple of checks are no longer skipped.

Does returning `TimeDeltaBlock(new_values, slice(len(new_values)))` work?

mroeschke commented:
Thanks @TomAugspurger; I was able to get the block placement to work when passing the slice from `self.mgr_locs.indexer`

jreback commented on pandas/tests/frame/test_timeseries.py:
```diff
@@ -57,6 +57,29 @@ def test_diff(self):
         assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize('axis', [0, 1])
+    @pytest.mark.parametrize('tz', [None, 'UTC'])
```
can we separate this into 2 tests, one for axis 0 and one for axis 1. The reason is that on axis 1 need to also test for *different* timezone diffs (which should fail). I think you are going to raise NotImplementedError, but this should really be a TypeError.

mroeschke replied:
`axis=1` currently doesn't work even for the same timezone:

```
In [2]: df = DataFrame({0: date_range('2010', freq='D', periods=2, tz='utc'),
   ...:                 1: date_range('2010', freq='D', periods=2, tz='utc')})
   ...:

In [3]: df.diff(axis=1) # incorrect, commented out the NotImplementedError
Out[3]:
       0      1
0    NaT    NaT
1 1 days 1 days

In [4]: df.diff(axis=0) # correct
Out[4]:
       0      1
0    NaT    NaT
1 1 days 1 days
```
since diff operates per block via apply. As you mentioned before this probably needs a whole new implementation in general, but that's why I thought of raising a `NotImplementedError` for `diff(axis=1)` (for now)

jreback commented on pandas/core/internals.py:
```diff
@@ -2905,6 +2905,18 @@ def shift(self, periods, axis=0, mgr=None):
                                            placement=self.mgr_locs)]
 
+    def diff(self, n, axis=0, mgr=None):
+        """1st discrete difference"""
```
can you make a full doc-string.

mroeschke committed:
Split test and add docstring

mroeschke commented:
Split test for axis=1 and axis=0, added docstring.

All green.

jreback commented on pandas/core/internals.py:
```diff
@@ -2905,6 +2905,34 @@ def shift(self, periods, axis=0, mgr=None):
         return [self.make_block_same_class(new_values,
                                            placement=self.mgr_locs)]
 
+    def diff(self, n, axis=0, mgr=None):
```
this should be in DatetimeBlock, puzzled why this works

mroeschke replied:
You had mentioned, https://github.com/pandas-dev/pandas/pull/19773#discussion_r169189270, that this should belong in DatetimeTZBlock. 

Additionally, DatetimeBlock can handle 2D data which can use the `algos.diff` method (dispatched via `Block.apply('diff', **kwargs)`) since that data can be converted to np.datetime64 cleanly. 

`DatetimeTZBlock` here overwrites the Block diff method that calls `algos.diff` but mimics the signature so everything is happy.

jreback replied:
yeah, ok for now. this whole diff bizness needs to be completely refactored.

jreback commented on pandas/core/internals.py:
```diff
@@ -2905,6 +2905,34 @@ def shift(self, periods, axis=0, mgr=None):
+            # function is invoked via apply
+            raise NotImplementedError
+        new_values = (self.values - self.shift(n, axis=axis)[0].values).asi8
+        # Reshape the new_values like how algos.diff does for timedelta data
```
can you add a blank line here

mroeschke committed:
add blank line

jreback commented:
thanks!
